File: Data Structures & Algorithms/generate-parentheses/submission-0.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        
        s = ""
        res = []


        def dfs(o , c ,s):
            if o == c == n :
                res.append(s)
                return
            if o < n:
                s += "("
                dfs(o+1 , c ,s)
                s = s[:-1]
            if c < o:
                s += ")"
                dfs(o , c+1 ,s)
                s = s[:-1]

            
        dfs(0,0 , s )

        return res
            


            

File: Data Structures & Algorithms/permutations-ii/submission-0.cpp
class Solution {
public:
    vector<vector<int>> res;
    set<vector<int>> tab;
    
    void dfs(vector<int>& nums ,vector<bool>& used ,vector<int>& cur){
        if(cur.size() == nums.size()){
            if(tab.find(cur) == tab.end()){
                res.push_back(cur);
                tab.insert(cur);
            }
            return;
        }
        

        for(int i= 0 ;i < nums.size(); i++){
            if(!used[i]){
                cur.push_back(nums[i]);
                used[i] = true;
                dfs(nums , used , cur);
                cur.pop_back();
                used[i]= false;
            }
        }
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<bool> used(nums.size() , false);
        
        vector<int> cur;
        dfs(nums , used , cur);
        return res;
        
    }
};